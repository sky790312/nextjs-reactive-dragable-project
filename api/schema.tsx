export interface PAGE {
  id: number
  name: string
  url: string
}

export interface BLOCK {
  id: number
  name: string
  x: number
  y: number
  o: number
  c: string
}
