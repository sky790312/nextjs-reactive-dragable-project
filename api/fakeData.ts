/**
 * pages should come from somewhere and get all url config.
 */
export const FAKE_PAGES = [
  {
    id: 1,
    name: 'page1',
    url: '/pages/1',
  },
  {
    id: 2,
    name: 'page2',
    url: '/pages/2',
  },
]

/**
 * items should come from api(maybe by page) and get x, y, o style.
 */
export const FAKE_BLOCKS = [
  {
    id: 1,
    name: 'Element 1-1',
    x: 10,
    y: 10,
    o: 1,
    c: '#008800',
  },
  {
    id: 2,
    name: 'Element 1-2',
    x: 60,
    y: 60,
    o: 0.5,
    c: '#008800',
  },
  {
    id: 3,
    name: 'Element 1-3',
    x: 110,
    y: 110,
    o: 1,
    c: '#008800',
  },
  {
    id: 4,
    name: 'Element 2-1',
    x: 30,
    y: 30,
    o: 1,
    c: '#008800',
  },
]
