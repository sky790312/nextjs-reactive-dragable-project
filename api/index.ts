import { FAKE_PAGES, FAKE_BLOCKS } from '@/api/fakeData'
import { PAGE, BLOCK } from '@/api/schema'

export const fetchPages = async () => {
  const data: PAGE[] = await FAKE_PAGES
  return data
}

export const fetchBlocks = async () => {
  const data: BLOCK[] = await FAKE_BLOCKS
  return data
}
