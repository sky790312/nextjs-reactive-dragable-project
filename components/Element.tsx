import React, { useState } from 'react'
import EditableItem from '@/components/EditableItem'
import { BLOCK } from '@/api/schema'
import { Dispatch } from 'redux'
import { blockActions } from '@/store/state.block'
import { useSelector, shallowEqual, useDispatch } from 'react-redux'
import { RootState } from '@/store/index'

interface Props {
  block: BLOCK
}

const Element: React.FC<Props> = React.memo(({ block }) => {
  const currentBlock = useSelector(
    (state: RootState) => state.block.currentBlock,
    shallowEqual,
  )
  const dispatch: Dispatch = useDispatch()
  const { setCurrentBlock } = blockActions
  const isCurrentBlock = block.id === currentBlock?.id
  const [blockName, setBlockName] = useState(block.name)

  const handleClick = () => {
    if (isCurrentBlock) {
      return
    }

    dispatch(setCurrentBlock(block))
  }

  const handleEditConfirm = (newName: string) => {
    setBlockName(newName)
  }

  return (
    <EditableItem
      key={block.id}
      text={blockName}
      shouldHighlight={isCurrentBlock}
      handleClick={handleClick}
      handleEditConfirm={handleEditConfirm}
    />
  )
})

export default Element
