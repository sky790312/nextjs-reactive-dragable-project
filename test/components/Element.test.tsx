import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import Element from '@/components/Element'
import { FAKE_BLOCKS } from '@/api/fakeData'

describe('Element component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      withProviders(<Element block={FAKE_BLOCKS[0]} />),
      {},
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
