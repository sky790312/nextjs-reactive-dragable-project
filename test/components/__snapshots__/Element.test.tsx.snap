// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="EditableItem__Text-apsj0b-0 jUHTNR"
    >
      Element 1-1
    </div>
  </div>
</DocumentFragment>
`;
